apps/src/applab/applab.js

```diff
@@ -1250,13 +1250,17 @@ function onInterfaceModeChange(mode) {
  */
 Applab.setWorkspaceMode = function (mode) {
   var designWorkspace = document.getElementById('designWorkspace');
-  designWorkspace.style.display = (ApplabInterfaceMode.DESIGN === mode) ? 'block' : 'none';
+  if (designWorkspace) {
+    designWorkspace.style.display = (ApplabInterfaceMode.DESIGN === mode) ? 'block' : 'none';
+  }
 
   var codeWorkspaceWrapper = document.getElementById('codeWorkspaceWrapper');
   codeWorkspaceWrapper.style.display = (ApplabInterfaceMode.CODE === mode) ? 'block' : 'none';
 
   var dataWorkspaceWrapper = document.getElementById('dataWorkspaceWrapper');
-  dataWorkspaceWrapper.style.display = (ApplabInterfaceMode.DATA === mode) ? 'block' : 'none';
+  if (dataWorkspaceWrapper) {
+    dataWorkspaceWrapper.style.display = (ApplabInterfaceMode.DATA === mode) ? 'block' : 'none';
+  }
```

Comment:
OK, I see now why this is messy. I've made the following changes:
* extract the workspaces so they are passed in by AppView / AppLabView / GameLabView
* move the logic for whether DataWorkspace is visible into DataWorkspace
* move DataWorkspace into `src/applab`

However I think it is better to render a large hidden div and then show/hide it, rather than to repeatedly add/remove it from the DOM. There is some discussion of the performance benefits in the comments here: http://stackoverflow.com/questions/24502898/show-or-hide-element-in-react-js

FWIW the structure of the Redux state is still a little weird, because applab-specific variables (such as hasDataMode) are mixed in with other state which is accessed by other apps.